File: app/src/main/java/com/example/nfcreader/MainActivity.java
package com.example.nfcreader;

import android.app.Activity;
import android.app.PendingIntent;
import android.content.Intent;
import android.content.IntentFilter;
import android.nfc.NdefMessage;
import android.nfc.NdefRecord;
import android.nfc.NfcAdapter;
import android.nfc.Tag;
import android.nfc.tech.Ndef;
import android.os.Bundle;
import android.os.Parcelable;
import android.util.Log;
import android.widget.TextView;
import android.widget.Toast;

import java.io.UnsupportedEncodingException;
import java.util.Arrays;

public class MainActivity extends Activity {

    public static final String TAG = "NfcDemo";

    private TextView mTextView;
    private NfcAdapter mNfcAdapter;
    private PendingIntent mPendingIntent;
    private IntentFilter[] mFilters;
    private String[][] mTechLists;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mTextView = findViewById(R.id.text_view_result);

        mNfcAdapter = NfcAdapter.getDefaultAdapter(this);

        if (mNfcAdapter == null) {
            Toast.makeText(this, "This device doesn't support NFC.", Toast.LENGTH_LONG).show();
            finish();
            return;
        }

        if (!mNfcAdapter.isEnabled()) {
            mTextView.setText("NFC is disabled.");
        } else {
            mTextView.setText("NFC is ready. Scan a tag.");
        }

        // Create a PendingIntent object so the Android system can populate it with the details of the tag when it is scanned.
        // PendingIntent.FLAG_MUTABLE is required for Android 12+
        mPendingIntent = PendingIntent.getActivity(this, 0,
                new Intent(this, getClass()).addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP),
                PendingIntent.FLAG_MUTABLE);

        // Setup an intent filter for all MIME based dispatches
        IntentFilter ndef = new IntentFilter(NfcAdapter.ACTION_NDEF_DISCOVERED);
        try {
            ndef.addDataType("*/*");
        } catch (IntentFilter.MalformedMimeTypeException e) {
            throw new RuntimeException("fail", e);
        }
        mFilters = new IntentFilter[] {
                ndef,
        };

        // Setup a tech list for all NfcF tags
        mTechLists = new String[][] { new String[] { android.nfc.tech.NfcF.class.getName() } };
    }

    @Override
    public void onResume() {
        super.onResume();
        if (mNfcAdapter != null) {
            mNfcAdapter.enableForegroundDispatch(this, mPendingIntent, mFilters, mTechLists);
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if (mNfcAdapter != null) {
            mNfcAdapter.disableForegroundDispatch(this);
        }
    }

    @Override
    public void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        Log.i(TAG, "onNewIntent");
        handleIntent(intent);
    }

    private void handleIntent(Intent intent) {
        String action = intent.getAction();
        if (NfcAdapter.ACTION_NDEF_DISCOVERED.equals(action) ||
            NfcAdapter.ACTION_TECH_DISCOVERED.equals(action) ||
            NfcAdapter.ACTION_TAG_DISCOVERED.equals(action)) {
            
            Tag tag = intent.getParcelableExtra(NfcAdapter.EXTRA_TAG);
            String tagId = bytesToHex(tag.getId());
            
            StringBuilder sb = new StringBuilder();
            sb.append("Tag Detected!\n");
            sb.append("ID: ").append(tagId).append("\n");
            sb.append("Tech: ").append(Arrays.toString(tag.getTechList())).append("\n");

            Parcelable[] rawMsgs = intent.getParcelableArrayExtra(NfcAdapter.EXTRA_NDEF_MESSAGES);
            if (rawMsgs != null) {
                NdefMessage[] msgs = new NdefMessage[rawMsgs.length];
                for (int i = 0; i < rawMsgs.length; i++) {
                    msgs[i] = (NdefMessage) rawMsgs[i];
                    sb.append("\nMessage ").append(i + 1).append(":\n");
                    for (NdefRecord record : msgs[i].getRecords()) {
                        sb.append(parseRecord(record)).append("\n");
                    }
                }
            } else {
                sb.append("\nNo NDEF messages found.");
            }
            
            mTextView.setText(sb.toString());
        }
    }

    private String parseRecord(NdefRecord record) {
        try {
            byte[] payload = record.getPayload();
            String textEncoding = ((payload[0] & 128) == 0) ? "UTF-8" : "UTF-16";
            int languageCodeLength = payload[0] & 0063;
            return new String(payload, languageCodeLength + 1, payload.length - languageCodeLength - 1, textEncoding);
        } catch (UnsupportedEncodingException e) {
            Log.e(TAG, "Unsupported Encoding", e);
        } catch (Exception e) {
            Log.e(TAG, "Error parsing record", e);
        }
        return "Unknown Record Type";
    }

    private String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02X", b));
        }
        return sb.toString();
    }
}

File: app/src/main/java/com/example/nfcreader/MainActivity.kt
package com.example.nfcreader

import android.app.PendingIntent
import android.content.Intent
import android.nfc.NdefMessage
import android.nfc.NfcAdapter
import android.nfc.Tag
import android.nfc.tech.IsoDep
import android.nfc.tech.MifareClassic
import android.nfc.tech.MifareUltralight
import android.nfc.tech.Ndef
import android.nfc.tech.NdefFormatable
import android.nfc.tech.NfcA
import android.nfc.tech.NfcB
import android.nfc.tech.NfcF
import android.nfc.tech.NfcV
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.nfcreader.databinding.ActivityMainBinding
import java.nio.charset.Charset

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private var nfcAdapter: NfcAdapter? = null
    private lateinit var pendingIntent: PendingIntent

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        nfcAdapter = NfcAdapter.getDefaultAdapter(this)

        if (nfcAdapter == null) {
            binding.tvStatus.text = getString(R.string.nfc_not_supported)
            binding.tvTagData.text = ""
            return
        }

        if (!nfcAdapter!!.isEnabled) {
            binding.tvStatus.text = getString(R.string.nfc_disabled)
            Toast.makeText(this, R.string.nfc_enable_prompt, Toast.LENGTH_LONG).show()
        } else {
            binding.tvStatus.text = getString(R.string.ready_to_scan)
        }

        // PendingIntent used for foreground dispatch — routes all NFC intents to this activity
        val intent = Intent(this, javaClass).addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
        pendingIntent = PendingIntent.getActivity(
            this, 0, intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_MUTABLE
        )

        // If the activity was started by an NFC intent, handle it immediately
        handleIntent(getIntent())
    }

    override fun onResume() {
        super.onResume()
        nfcAdapter?.enableForegroundDispatch(this, pendingIntent, null, null)
    }

    override fun onPause() {
        super.onPause()
        nfcAdapter?.disableForegroundDispatch(this)
    }

    /** Called when a new Intent arrives while the activity is running (foreground dispatch). */
    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        handleIntent(intent)
    }

    // -------------------------------------------------------------------------
    // NFC processing
    // -------------------------------------------------------------------------

    private fun handleIntent(intent: Intent) {
        val action = intent.action ?: return

        if (action != NfcAdapter.ACTION_NDEF_DISCOVERED &&
            action != NfcAdapter.ACTION_TECH_DISCOVERED &&
            action != NfcAdapter.ACTION_TAG_DISCOVERED
        ) return

        val tag: Tag = intent.getParcelableExtra(NfcAdapter.EXTRA_TAG) ?: return

        val sb = StringBuilder()

        // --- Tag ID ---
        sb.appendLine("Tag ID: ${tag.id.toHex()}")

        // --- Technologies ---
        sb.appendLine("Technologies: ${tag.techList.joinToString { it.substringAfterLast('.') }}")
        sb.appendLine()

        // --- NDEF payload (if available) ---
        val ndef = Ndef.get(tag)
        if (ndef != null) {
            try {
                ndef.connect()
                val ndefMessage: NdefMessage? = ndef.ndefMessage
                if (ndefMessage != null) {
                    sb.appendLine("NDEF Records (${ndefMessage.records.size}):")
                    ndefMessage.records.forEachIndexed { index, record ->
                        sb.appendLine("  Record ${index + 1}:")
                        sb.appendLine("    TNF: ${record.tnf}")
                        sb.appendLine("    Type: ${String(record.type, Charset.forName("US-ASCII"))}")
                        val payload = record.payload
                        // For well-known text records, skip the language-code prefix
                        val text = if (record.tnf == android.nfc.NdefRecord.TNF_WELL_KNOWN &&
                            record.type.contentEquals(android.nfc.NdefRecord.RTD_TEXT) &&
                            payload.isNotEmpty()
                        ) {
                            val languageCodeLength = payload[0].toInt() and 0x3F
                            val textStart = languageCodeLength + 1
                            if (textStart <= payload.size) {
                                String(payload, textStart, payload.size - textStart, Charsets.UTF_8)
                            } else {
                                String(payload, Charsets.UTF_8)
                            }
                        } else {
                            String(payload, Charsets.UTF_8)
                        }
                        sb.appendLine("    Payload: $text")
                    }
                } else {
                    sb.appendLine("NDEF: empty message")
                }
            } catch (e: Exception) {
                sb.appendLine("NDEF read error: ${e.message}")
            } finally {
                try { ndef.close() } catch (_: Exception) {}
            }
        } else {
            sb.appendLine("Tag is not NDEF formatted.")
        }

        // --- Additional tech info ---
        sb.appendLine()
        sb.appendLine("--- Tag Details ---")

        NfcA.get(tag)?.let { tech ->
            try {
                tech.connect()
                sb.appendLine("NFC-A  ATQA: ${tech.atqa.toHex()}  SAK: ${tech.sak}")
            } catch (_: Exception) {} finally { try { tech.close() } catch (_: Exception) {} }
        }

        NfcB.get(tag)?.let { tech ->
            try {
                tech.connect()
                sb.appendLine("NFC-B  App Data: ${tech.applicationData.toHex()}")
            } catch (_: Exception) {} finally { try { tech.close() } catch (_: Exception) {} }
        }

        NfcF.get(tag)?.let { tech ->
            try {
                tech.connect()
                sb.appendLine("NFC-F  Manufacturer: ${tech.manufacturer.toHex()}")
            } catch (_: Exception) {} finally { try { tech.close() } catch (_: Exception) {} }
        }

        NfcV.get(tag)?.let { tech ->
            try {
                tech.connect()
                sb.appendLine("NFC-V  DSF ID: ${tech.dsfId}  Response flags: ${tech.responseFlags}")
            } catch (_: Exception) {} finally { try { tech.close() } catch (_: Exception) {} }
        }

        MifareClassic.get(tag)?.let { tech ->
            try {
                tech.connect()
                sb.appendLine("MIFARE Classic  Size: ${tech.size} bytes  Sectors: ${tech.sectorCount}")
            } catch (_: Exception) {} finally { try { tech.close() } catch (_: Exception) {} }
        }

        MifareUltralight.get(tag)?.let { tech ->
            try {
                tech.connect()
                sb.appendLine("MIFARE Ultralight  Type: ${tech.type}")
            } catch (_: Exception) {} finally { try { tech.close() } catch (_: Exception) {} }
        }

        IsoDep.get(tag)?.let { tech ->
            try {
                tech.connect()
                sb.appendLine("ISO-DEP  Hi Layer Response: ${tech.hiLayerResponse?.toHex() ?: "N/A"}")
            } catch (_: Exception) {} finally { try { tech.close() } catch (_: Exception) {} }
        }

        NdefFormatable.get(tag)?.let {
            sb.appendLine("Tag is NDEF-formatable (currently unformatted).")
        }

        binding.tvTagData.text = sb.toString()
        binding.tvStatus.text = getString(R.string.tag_detected)
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private fun ByteArray.toHex(): String =
        joinToString(separator = "") { byte -> "%02X".format(byte) }
}
